python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py

```diff
@@ -1063,22 +1069,10 @@ def equivalent_to_stored_evaluation(
             and stored_evaluation.num_requested == 0
             and stored_evaluation.num_discarded == 0
             and stored_evaluation.num_skipped == self.num_skipped
-            and len(sorted_results) == len(sorted_stored_results)
-            and (
-                # first is a quick check for the equality of the string representations of the
-                # partition subsets
-                sorted_results == [tuple(x) for x in sorted_stored_results]
-                # however, not all identical partition subsets are serialized to the same string,
-                # so sometimes we need to deserialize the keys to be sure
-                or [
-                    self._deserialize_rule_evaluation_result(re, ss, asset_graph)
-                    for re, ss in sorted_results
-                ]
-                == [
-                    self._deserialize_rule_evaluation_result(re, ss, asset_graph)
-                    for re, ss in sorted_stored_results
-                ]
-            )
+            # when rule evaluation results are deserialized from json, they are lists instead of
+            # tuples, so we must convert them before comparing
+            and sorted(self.partition_subsets_by_condition)
+            == sorted([tuple(x) for x in stored_evaluation.partition_subsets_by_condition])
```

Comment:
going to come back and understand this, is the passed in object not the same as the calling object?

Reply:
it's mostly a quirk of our namedtuple serdes -- basically any time you have a `tuple` type as one of your fields, we break the invariant `obj == deserialize(serialize(obj))` because tuples get deserialized as lists (because json doesn't have a concept of tuples)

Reply:
Ok, yikes. @alangenfeld as unofficial Vice President of serdes was that on your radar? seems like a bit of a landmine, I am a bit surprised that it doesn't raise an exception when you try to serialize an unnamed tuple.

Should we do this translation on deserialization in a NamedTupleSerializer for the object rather than on a case-by-case basis when somebody is accessing partition_subsets_by_condition and it might have come from serialized JSON?

Not blocking since that is something that can be done later, but it seems like a good idea so that you don't have to think about two different types every time you are passing around one of these objects

Reply:
Yep agreed, and I found out the hard way about this one -- the other option would just to put this in the `__new__` method

Reply:
was aware of json tuple issue but not how it could manifest in whitelist for serdes fields. Wonder if there are any other tuple fields past this one. We could raise or add support like we do with `set` / `frozenset`

Reply:
probably worthwhile to add `__new__` and do runtime validation on `AutoMaterializeAssetEvaluation`

Reply:
https://github.com/dagster-io/dagster/pull/17556